### PR TITLE
BUG: local_binary_pattern platform mismatch fix

### LINKED ISSUE #8198: BUG: local_binary_pattern platform-specific behavior
While working on a computer vision machine learning pipeline (with an underexplored class of estimators: https://github.com/scikit-learn/scikit-learn/pull/33215 and https://github.com/lanl/GFDL) that operates on the Fashion MNIST dataset, we noticed that we were getting slightly different results between MacOS and Linux. We first thought this was a scaling issue (https://github.com/scikit-learn/scikit-learn/issues/34082), but upon closer inspection, I'm now pretty confident this is an issue in `local_binary_pattern()` producing "off by 1" numerical discrepancies between MacOS ARM and Linux x86_64 for the same image input data.

A sample regression test that passes on MacOS ARM and fails on x86_64 Linux is pasted below, and also available in feature branch `treddy_linux_mac_LBP_mismatch` on my fork of `scikit-image`. This was carefully dissected out of a much larger processing pipeline. I don't actually know offhand if Mac or Linux is "correct" here, or if we have a good reference implementation to check (i.e., OpenCV or something else?).

The overall impact was subtle but real, with this type of problem occurring for several images in F-MNIST, and our final design matrix differing between the two platforms by 826 / 1080000 elements (0.0765%). 

```diff
--- a/tests/skimage/feature/test_texture.py
+++ b/tests/skimage/feature/test_texture.py
@@ -262,6 +262,22 @@ class TestLBP:
         )
         np.testing.assert_array_equal(lbp, ref)
 
+
+    def test_linux_mac_portability(self):
+        try:
+            import sklearn
+        except ImportError:
+            pytest.skip("test requires sklearn")
+        from sklearn.datasets import fetch_openml
+        dataset = fetch_openml("Fashion-MNIST")
+        X = dataset.data
+        images = X.to_numpy().reshape(-1, 28, 28)
+        image = images[135]
+        LBP = local_binary_pattern(image, P=16, R=2, method="uniform")
+        actual = LBP[-8][11:15]
+        np.testing.assert_array_equal(actual, [11, 9, 11, 13])
+
+

```

### PR DESCRIPTION
**Pull request by tylerjereddy**
Fixes gh-8198.

* Side-by-side debug printing between MacOS and Linux for the Cython code underlying the regression test added here indicated a first point of deviation between platforms that was related to a floating point comparison with `0`. For the image associated with this regression test, the delta values were `-2.842170943040401e-14` and `0.0` (based on what gets printed at least, after temporarily removing the `nogil` block to debug).

* The full testsuite seem to pass locally on Linux x86_64 and Mac OS ARM on this feature branch, and the test seems to fail before and pass after the patch.

AI disclosure: I used Claude Sonnet 4.5 on our internal AI portal to probe the source code of the `_local_binary_pattern()` Cython function and rank the most likely locations for platform-specific differences. I've reproduced that ranking below the fold. That helped me narrow down the debug print locations, which I then ran on both platforms and diffed the verbose output to identify the first point of platform-specific deviation.

<details>

image

</details>

### CODE REVIEW

File: src/skimage/feature/_texture.pyx [hunk 1]
```diff
@@ -124,6 +124,7 @@ def _local_binary_pattern(cnp.float64_t[:, ::1] image,
 
     # pre-allocate arrays for computation
     cdef cnp.float64_t[::1] texture = np.zeros(P, dtype=np.float64)
+    cdef cnp.float64_t epsilon = 1e-12
```
Comment (tylerjereddy) [inline src/skimage/feature/_texture.pyx:127]: Perhaps it might be more suitable to allow the user to control the epsilon/threshold via a new argument to `local_binary_pattern`? Unless there's a great rule of thumb for what threshold would make sense?

Comment (tylerjereddy) [reply src/skimage/feature/_texture.pyx:127]: I believe all `method`s may be affected by this issue in principle, based on the control flow in the Cython function.

File: tests/skimage/feature/test_texture.py [hunk 2]
```diff
@@ -262,6 +262,22 @@ def test_default(self):
...
+        except ImportError:
+            pytest.skip("test requires sklearn")
+        from sklearn.datasets import fetch_openml
+        dataset = fetch_openml("Fashion-MNIST")
+        X = dataset.data
+        images = X.to_numpy().reshape(-1, 28, 28)
+        image = images[135]
+        LBP = local_binary_pattern(image, P=16, R=2, method="uniform")
+        actual = LBP[-8][11:15]
+        np.testing.assert_array_equal(actual, [11, 9, 11, 13])
```
Comment (tylerjereddy) [inline tests/skimage/feature/test_texture.py:278]: This test has the advantage that it is based on a real-world scenario where it caused a problem in a machine learning pipeline we have.

It has the drawback of depending on `sklearn` (`git grep` suggests you have a few other tests that optionally run if `sklearn` is present) and loading in a decently-sized dataset, and `fetch_openml` can flake on occasion depending on maintenance of servers I think, based on our experience with in house test suites.

The other drawback is that the expected value is not "ground truth"/based on some "oracle--" it is just the common result that both platforms agree upon when switching to usage of a threshold value.

It may also be useful to see if this branch fully resolves the in house workflow problem, since this is just one of many sample images that caused problems, and the choice of `epsilon` may need to change a bit for different images for platforms/floating point math to agree.

One other consideration is if it might be possible to avoid source changes and use build flags that allow the floating point math to be a bit more stable--seems like a stretch?

File: tests/skimage/feature/test_texture.py [hunk 3]
```diff
@@ -262,6 +262,22 @@ def test_default(self):
...
+
+    def test_gh_8198(self):
+        try:
+            import sklearn
+        except ImportError:
+            pytest.skip("test requires sklearn")
+        from sklearn.datasets import fetch_openml
+        dataset = fetch_openml("Fashion-MNIST")
+        X = dataset.data
+        images = X.to_numpy().reshape(-1, 28, 28)
```
[the remark this answers is not in the document]
Comment (tylerjereddy) [reply tests/skimage/feature/test_texture.py:274]: > A more robust alternative would be to construct a minimal synthetic image (or store a tiny NumPy fixture)

I mostly agree with the AI here as my other comments indicate--but generating a suitable reproducer may be a bit of work on cutting down. We might be able to use `hypothesis`, although it requires probing for differences across platforms, which is annoying.

**Comment by stefanv** [comment]
Hi Tyler, thanks for flagging this. I'm wondering: is a -2.842170943040401e-14 delta worth changing any code for? That seems within the realm of possible floating point differences between platforms.

I haven't read the code carefully, but I'm a bit concerned about adding a futz factor to a check for positivity. What should that factor be, and why is it valid? The futz factor is larger than the discrepancy witnessed, so isn't it likely that it will introduce greater changes itself?

Totally fine to use AI to hunt down these types of bugs. I hope you can help us understand the fix better too.

File: tests/skimage/feature/test_texture.py (hunk 2 above)
Comment (stefanv) [reply tests/skimage/feature/test_texture.py:278]: I suspect that, with some random searching, we can also find a test matrix to replicate the result. But let's talk about whether it is worth fixing in the first place (main thread comment).

**Comment by tylerjereddy** [comment]
It looks like the problem is happening upstream in the inlined `bilinear_interpolation()`, in the texture value assignments, based on more debug printing (below the fold). While those differences look very small, the end result is around 800 differences in the design matrix for F-MNIST, and a platform estimator accuracy difference around 1e-4 overall (as noted in the related issues).

I'm pretty sure that the general pattern `(a - b) >= 0` is a bit risky in floating point math across platforms. I'll try to check now/soon-ish (it is a pain to check), but I'm pretty sure that using some kind of epsilon threshold will allow the platforms to match, probably with some risk of eventual "algorithmic blunting" if it is set too high in magnitude. What the threshold should be I don't know (it can be debated perhaps).

<details>

```diff
--- a/mac_out.txt
+++ b/linux_out.txt
@@ -474,7 +474,7 @@ texture[i]: 0.0
 image[r, c]: 0.0
 texture[i]: 103.3339507512
 image[r, c]: 0.0
-texture[i]: 127.85821865259997
+texture[i]: 127.85821865259996
 image[r, c]: 0.0
 texture[i]: 127.3414033856
 image[r, c]: 0.0
@@ -552,7 +552,7 @@ texture[i]: 31.67780924880002
 image[r, c]: 0.0
 texture[i]: 158.0
 image[r, c]: 0.0
-texture[i]: 204.9324118832
+texture[i]: 204.93241188320002
```

</details>

**Comment by tylerjereddy** [comment]
I performed an `epsilon` "titration" with F-MNIST using the sample code below the fold on each platform with an `skimage` branch similar to this one, then comparing the saved arrays with `assert_allclose` to determine the number of design matrix mismatches. The results are summarized in the table below.

I don't think this is particularly surprising. We could try the titration on a few other datasets like regular MNIST and E-MNIST perhaps. As noted above though, the greatest flexibility may come from allowing the user to modulate the threshold via an exposed argument, which might default to a value we find generally consistent across a wide range of datasets (or you could default to the current behavior with a documented caveat that using a non-zero threshold is recommended for many common datasets). A few choices to consider perhaps.

| epsilon | `assert_allclose` mismatch count Linux x86_64 vs. Mac OS ARM|
| :--- | :---: | 
| `1e-12` | 0 | 
|  `4e-14` | 0 |
| `3e-14` | 0 |
| `2.8422e-14` |  0 |
| `2.8421e-14` | 0 |
| `2.84e-14` | 684 | 
| `2.5e-14` | 684 |  
| `2e-14` | 684| 
| `1e-14` | 750|
| `1e-15` | 826 |
 


<details>

```python
import platform
import numpy as np
from skimage.feature import local_binary_pattern
from tqdm import tqdm
from sklearn.datasets import fetch_openml


def main():
    dataset = fetch_openml("Fashion-MNIST")
    X, y = dataset.data, dataset.target
    y = y.astype(np.int8)
    images = X.to_numpy().reshape(-1, 28, 28)
    radius = 2
    n_points = 8 * radius
    n_bins = n_points + 2
    hog_features = []
    for image in tqdm(images):
        new_feats = []
        LBP = local_binary_pattern(image,
                                   P=n_points,
                                   R=radius,
                                   method="uniform")
        lbp_hist, _ = np.histogram(LBP.ravel(),
                                   bins=n_bins,
                                   range=(0, n_bins))
        new_feats.append(lbp_hist)
        hog_features.append(np.atleast_2d(new_feats))
    X = np.asarray(hog_features).reshape(X.shape[0], -1)
    X_train = X[:60_000]
    sys_name = platform.system()
    np.save(f"{sys_name}_X_train.npy", X_train)


if __name__ == "__main__":
    main()

```

</details>

**Comment by michaelbratsch** [comment]
To me it seems that changing the threshold from zero to epsilon does not fully resolve the platform-specific differences. I think it should still be possible to get different outputs for artificial examples.

An alternative approach is to modify the evaluation in the bilinear interpolation to get improved results for equal endpoints: https://github.com/scikit-image/scikit-image/pull/8207.

**Comment by tylerjereddy** [comment]
Ok, if you normally operate by superseding PRs instead of suggesting improvements fair enough.

I still don't think you should be performing an exact floating point subtraction comparison to `0` in your algorithm, but if I've convinced the team that there's a problematic bug I suppose I've accomplished what is needed here.

**Comment by michaelbratsch** [comment]
> Ok, if you normally operate by superseding PRs instead of suggesting improvements fair enough.

Sorry for being so harsh. I already started working on the other PR before this one. Nevertheless I should have coordinated with you instead of just opening a new one. Criticism is noted.

> I still don't think you should be performing an exact floating point subtraction comparison to `0` in your algorithm, but if I've convinced the team that there's a problematic bug I suppose I've accomplished what is needed here.

I know that such epsilon checks are often a good approach when dealing with floating point numbers. To me it's currently not fully clear if we are solving what we are intending to do and it gives reproducibility across platforms. Will have to think about it a bit more.

**Comment by michaelbratsch** [comment]
> I performed an `epsilon` "titration" with F-MNIST using the sample code below the fold on each platform with an `skimage` branch similar to this one, then comparing the saved arrays with `assert_allclose` to determine the number of design matrix mismatches. The results are summarized in the table below.
> 
> I don't think this is particularly surprising. We could try the titration on a few other datasets like regular MNIST and E-MNIST perhaps. As noted above though, the greatest flexibility may come from allowing the user to modulate the threshold via an exposed argument, which might default to a value we find generally consistent across a wide range of datasets (or you could default to the current behavior with a documented caveat that using a non-zero threshold is recommended for many common datasets). A few choices to consider perhaps.
> 
> epsilon	`assert_allclose` mismatch count Linux x86_64 vs. Mac OS ARM
> `1e-12`	0
> `4e-14`	0
> `3e-14`	0
> `2.8422e-14`	0
> `2.8421e-14`	0
> `2.84e-14`	684
> `2.5e-14`	684
> `2e-14`	684
> `1e-14`	750
> `1e-15`	826
> Details

To provide a reasonable epsilon value that fits as many use cases as possible, one could scale it with the input values so that it depends on whether we use `uint8` or `uint16` input images. I expect the dataset above to be always `uint8`, right?

So something similar to

```python
from libc.float cimport DBL_EPSILON
from libc.math cimport fabs
cdef cnp.float64_t rtol = 8 * DBL_EPSILON
if texture[i] - image[r, c] >= -rtol * fabs(image[r, c])
```

**Comment by michaelbratsch** [comment]
> I still don't think you should be performing an exact floating point subtraction comparison to `0` in your algorithm, but if I've convinced the team that there's a problematic bug I suppose I've accomplished what is needed here.

The epsilon of the floating point subtraction comparison serves two purposes:
* Remove the platform dependency (for integer images)
  * for float images we still have the platform dependency problem (but there is already a note) https://github.com/scikit-image/scikit-image/blob/8ce7e0942360329fef15939e1b3e5c44fe941edc/src/skimage/feature/texture.py#L386
* Change the way float-near ties are being classified (for float images)
  * could be seen as an additional feature which is unrelated to the platform dependency

So to make it work, we need to find an epsilon that serves both purposes well which might be not so easy. Making the epsilon configurable is one option but would probably just mean that the user doesn't know how to set.

**Comment by matthew-brett** [comment]
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

Read on for the rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
